File: go/me_test.go
package realmid

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
)

// meServer wires the platform-token bootstrap plus the caller's handlers.
func meServer(t *testing.T, handlers map[string]http.HandlerFunc) *httptest.Server {
	t.Helper()
	mux := http.NewServeMux()
	mintPlatformToken(mux)
	for p, h := range handlers {
		mux.HandleFunc(p, h)
	}
	return httptest.NewServer(mux)
}

func TestMe_ChooseTenantSendsKeptTenantAndDecodesReleased(t *testing.T) {
	var gotBody map[string]any
	var gotAuth string
	srv := meServer(t, map[string]http.HandlerFunc{
		"/me/tenant-choice": func(w http.ResponseWriter, r *http.Request) {
			if r.Method != "POST" {
				t.Errorf("method=%s", r.Method)
			}
			gotAuth = r.Header.Get("Authorization")
			buf, _ := io.ReadAll(r.Body)
			_ = json.Unmarshal(buf, &gotBody)
			_ = json.NewEncoder(w).Encode(map[string]any{
				"tenant_id": "t-keep", "status": "chosen", "released": 2,
			})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	out, err := r.Me.ChooseTenant(context.Background(), TenantChoiceRequest{
		MeAuth:   MeAuth{UserBearer: "user-jwt"},
		TenantID: "t-keep",
	})
	if err != nil {
		t.Fatalf("choose: %v", err)
	}
	// The body names the membership to KEEP, not the ones released.
	if gotBody["tenant_id"] != "t-keep" {
		t.Errorf("body tenant_id=%#v", gotBody["tenant_id"])
	}
	if gotAuth != "Bearer user-jwt" {
		t.Errorf("authorization=%q, want the user bearer in direct mode", gotAuth)
	}
	if out.Status != "chosen" || out.Released != 2 || out.TenantID != "t-keep" {
		t.Errorf("result=%+v", out)
	}
}

func TestMe_BFFModeSendsUserTokenBesidePlatformBearer(t *testing.T) {
	var gotAuth, gotUserToken string
	srv := meServer(t, map[string]http.HandlerFunc{
		"/me/tenant-choice": func(w http.ResponseWriter, r *http.Request) {
			gotAuth = r.Header.Get("Authorization")
			gotUserToken = r.Header.Get("X-User-Token")
			_ = json.NewEncoder(w).Encode(map[string]any{"tenant_id": "t1", "status": "chosen"})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	if _, err := r.Me.ChooseTenant(context.Background(), TenantChoiceRequest{
		MeAuth:   MeAuth{UserToken: "verified-jwt"},
		TenantID: "t1",
	}); err != nil {
		t.Fatalf("choose: %v", err)
	}
	// The platform token stays the bearer; the user JWT is additive. A BARE
	// user id is not an identity (issuer v0.66.0), so the verified token is
	// the only thing that names the caller.
	if gotAuth != "Bearer ptok" {
		t.Errorf("authorization=%q, want the platform token", gotAuth)
	}
	if gotUserToken != "verified-jwt" {
		t.Errorf("x-user-token=%q", gotUserToken)
	}
}

func TestMe_CtxUserTokenIsUsedWhenNoneOnTheRequest(t *testing.T) {
	var gotUserToken string
	srv := meServer(t, map[string]http.HandlerFunc{
		"/me/memberships/t1/leave": func(w http.ResponseWriter, r *http.Request) {
			gotUserToken = r.Header.Get("X-User-Token")
			_ = json.NewEncoder(w).Encode(map[string]any{"tenant_id": "t1", "status": "left"})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	ctx := WithUserToken(context.Background(), "ctx-jwt")
	if _, err := r.Me.LeaveMembership(ctx, MembershipRequest{TenantID: "t1"}); err != nil {
		t.Fatalf("leave: %v", err)
	}
	if gotUserToken != "ctx-jwt" {
		t.Errorf("x-user-token=%q, want the ctx-stashed token", gotUserToken)
	}
}

func TestMe_RejectAndLeaveHitTheirOwnRoutes(t *testing.T) {
	hit := map[string]bool{}
	srv := meServer(t, map[string]http.HandlerFunc{
		"/me/invitations/t%20one/reject": func(w http.ResponseWriter, r *http.Request) {
			hit["reject"] = true
			// No request body — the path and the session say everything.
			buf, _ := io.ReadAll(r.Body)
			if len(buf) != 0 {
				t.Errorf("reject body=%q, want empty", buf)
			}
			_ = json.NewEncoder(w).Encode(map[string]any{"tenant_id": "t one", "status": "rejected"})
		},
		"/me/memberships/t1/leave": func(w http.ResponseWriter, _ *http.Request) {
			hit["leave"] = true
			_ = json.NewEncoder(w).Encode(map[string]any{"tenant_id": "t1", "status": "left"})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	// A tenant id is a path SEGMENT and must be escaped, not interpolated raw.
	rej, err := r.Me.RejectInvitation(context.Background(), MembershipRequest{
		MeAuth: MeAuth{UserBearer: "u"}, TenantID: "t one",
	})
	if err != nil {
		t.Fatalf("reject: %v", err)
	}
	if rej.Status != "rejected" {
		t.Errorf("reject status=%q", rej.Status)
	}
	left, err := r.Me.LeaveMembership(context.Background(), MembershipRequest{
		MeAuth: MeAuth{UserBearer: "u"}, TenantID: "t1",
	})
	if err != nil {
		t.Fatalf("leave: %v", err)
	}
	if left.Status != "left" {
		t.Errorf("leave status=%q", left.Status)
	}
	if !hit["reject"] || !hit["leave"] {
		t.Errorf("routes hit=%v", hit)
	}
}

func TestMe_OwnerCannotBeRevokedSurfacesAsRealmError(t *testing.T) {
	srv := meServer(t, map[string]http.HandlerFunc{
		"/me/tenant-choice": func(w http.ResponseWriter, _ *http.Request) {
			w.WriteHeader(http.StatusConflict)
			_ = json.NewEncoder(w).Encode(map[string]any{
				"error": map[string]any{
					"code":    "owner_cannot_be_revoked",
					"message": "transfer ownership first",
				},
			})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	_, err := r.Me.ChooseTenant(context.Background(), TenantChoiceRequest{
		MeAuth: MeAuth{UserBearer: "u"}, TenantID: "t1",
	})
	if err == nil {
		t.Fatal("want an error")
	}
	re, ok := err.(*RealmError)
	if !ok {
		t.Fatalf("err type %T", err)
	}
	if re.Code != "owner_cannot_be_revoked" {
		t.Errorf("code=%q", re.Code)
	}
}

func TestAuth_LoginDecodesTenantChoicePicker(t *testing.T) {
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/auth/login": func(w http.ResponseWriter, _ *http.Request) {
			_ = json.NewEncoder(w).Encode(map[string]any{
				// The login SUCCEEDS: the picker rides alongside real tokens.
				"access_token":  "atok",
				"refresh_token": "rtok",
				"expires_in":    900,
				"user":          map[string]any{"id": "u1"},
				"tenants":       []any{},

				"tenant_choice_required": true,
				"tenant_choices": []any{
					map[string]any{"tenant_id": "t1", "display_name": "Acme", "is_owner": false},
					map[string]any{"tenant_id": "t2", "display_name": "Mine", "is_owner": true},
				},
			})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	s, err := r.Auth.Login(context.Background(), LoginRequest{Method: LoginGoogle, ProviderToken: "tok"})
	if err != nil {
		t.Fatalf("login: %v", err)
	}
	if s.AccessToken != "atok" {
		t.Errorf("access_token=%q — the picker is a prompt, not an auth failure", s.AccessToken)
	}
	if !s.TenantChoiceRequired {
		t.Error("TenantChoiceRequired=false")
	}
	if len(s.TenantChoices) != 2 {
		t.Fatalf("choices=%d", len(s.TenantChoices))
	}
	if s.TenantChoices[0].TenantID != "t1" || s.TenantChoices[0].DisplayName != "Acme" || s.TenantChoices[0].IsOwner {
		t.Errorf("choice[0]=%+v", s.TenantChoices[0])
	}
	// An owned membership cannot be given up — the client must not offer it.
	if !s.TenantChoices[1].IsOwner {
		t.Errorf("choice[1]=%+v, want is_owner", s.TenantChoices[1])
	}
}

func TestAuth_LoginWithoutPickerLeavesTheFieldsZero(t *testing.T) {
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/auth/login": func(w http.ResponseWriter, _ *http.Request) {
			_ = json.NewEncoder(w).Encode(map[string]any{
				"access_token": "atok", "refresh_token": "rtok", "expires_in": 900,
				"user": map[string]any{"id": "u1"}, "tenants": []any{},
			})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	s, err := r.Auth.Login(context.Background(), LoginRequest{Method: LoginGoogle, ProviderToken: "tok"})
	if err != nil {
		t.Fatalf("login: %v", err)
	}
	if s.TenantChoiceRequired || s.TenantChoices != nil {
		t.Errorf("picker leaked into an ordinary login: %+v / %+v", s.TenantChoiceRequired, s.TenantChoices)
	}
}

func TestConfig_GetSurfacesPendingReconciliationBesideConfig(t *testing.T) {
	srv := meServer(t, map[string]http.HandlerFunc{
		"/platforms/" + testRealmID + "/config": func(w http.ResponseWriter, _ *http.Request) {
			_ = json.NewEncoder(w).Encode(map[string]any{
				"id":     testRealmID,
				"config": map[string]any{"single_tenant_membership": true},
				// Derived, read-only, and deliberately OUTSIDE config.
				"single_tenant_pending_reconciliation": 3,
			})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	out, err := r.Config.Get(context.Background())
	if err != nil {
		t.Fatalf("get: %v", err)
	}
	if out.Config["single_tenant_membership"] != true {
		t.Errorf("single_tenant_membership=%#v", out.Config["single_tenant_membership"])
	}
	if _, leaked := out.Config["single_tenant_pending_reconciliation"]; leaked {
		t.Error("the derived count must not appear inside the settable config bag")
	}
	if out.SingleTenantPendingReconciliation == nil {
		t.Fatal("SingleTenantPendingReconciliation=nil, want 3")
	}
	if *out.SingleTenantPendingReconciliation != 3 {
		t.Errorf("pending=%d", *out.SingleTenantPendingReconciliation)
	}
}

func TestConfig_PendingReconciliationIsNilWhenTheRuleIsOff(t *testing.T) {
	srv := meServer(t, map[string]http.HandlerFunc{
		"/platforms/" + testRealmID + "/config": func(w http.ResponseWriter, _ *http.Request) {
			_ = json.NewEncoder(w).Encode(map[string]any{
				"id":     testRealmID,
				"config": map[string]any{"single_tenant_membership": false},
			})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	out, err := r.Config.Get(context.Background())
	if err != nil {
		t.Fatalf("get: %v", err)
	}
	// nil, not 0: "not reported" and "reported as fully drained" are different
	// facts and a caller rendering the number must be able to tell them apart.
	if out.SingleTenantPendingReconciliation != nil {
		t.Errorf("pending=%v, want nil", *out.SingleTenantPendingReconciliation)
	}
}

File: ts/src/me.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { createRealm } from "./realm.js";
import { RealmError } from "./errors.js";

interface Captured {
  method: string;
  url: string;
  body?: unknown;
  headers: Record<string, string>;
}

/** Serves the platform-token bootstrap, then defers to `handler`. */
function mkFetch(handler: (req: Captured) => Response): typeof fetch {
  return (async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input.toString();
    const method = init?.method ?? "GET";
    if (url.endsWith("/auth/login")) {
      return new Response(JSON.stringify({
        status: "ok", subject_type: "platform", refresh_token: "rtok", access_token: "pt_x", expires_in: 300,
      }), { status: 200, headers: { "content-type": "application/json" } });
    }
    let body: unknown;
    if (typeof init?.body === "string" && init.body.length > 0) {
      try { body = JSON.parse(init.body); } catch { body = init.body; }
    }
    const headers: Record<string, string> = {};
    for (const [k, v] of Object.entries((init?.headers ?? {}) as Record<string, string>)) {
      headers[k.toLowerCase()] = v;
    }
    return handler({ method, url, body, headers });
  }) as typeof fetch;
}

function json(payload: unknown, status = 200): Response {
  return new Response(JSON.stringify(payload), {
    status, headers: { "content-type": "application/json" },
  });
}

function realmWith(fetch: typeof fetch) {
  return createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
}

test("me.chooseTenant: posts the KEPT tenant and decodes released", async () => {
  let seen: Captured | undefined;
  const realm = realmWith(mkFetch((req) => {
    seen = req;
    return json({ tenant_id: "t-keep", status: "chosen", released: 2 });
  }));
  const out = await realm.me.chooseTenant({ tenantId: "t-keep", userBearer: "user-jwt" });
  assert.equal(seen?.method, "POST");
  assert.match(seen!.url, /\/me\/tenant-choice$/);
  assert.deepEqual(seen?.body, { tenant_id: "t-keep" });
  // Direct mode: the user's own JWT is the wire bearer.
  assert.equal(seen?.headers["authorization"], "Bearer user-jwt");
  assert.deepEqual(out, { tenantId: "t-keep", status: "chosen", released: 2 });
});

test("me: BFF mode sends X-User-Token beside the platform bearer", async () => {
  let seen: Captured | undefined;
  const realm = realmWith(mkFetch((req) => {
    seen = req;
    return json({ tenant_id: "t1", status: "chosen", released: 1 });
  }));
  await realm.me.chooseTenant({ tenantId: "t1", userToken: "verified-jwt" });
  // The platform token stays the bearer; the verified user JWT is additive.
  // A bare user id is not an identity (issuer v0.66.0), so this header is the
  // only thing naming the caller.
  assert.equal(seen?.headers["authorization"], "Bearer pt_x");
  assert.equal(seen?.headers["x-user-token"], "verified-jwt");
});

test("me.rejectInvitation / me.leave: own routes, no body, escaped tenant id", async () => {
  const hits: string[] = [];
  const realm = realmWith(mkFetch((req) => {
    hits.push(req.url);
    assert.equal(req.body, undefined, "these routes take no body");
    return json({ tenant_id: "t1", status: req.url.includes("reject") ? "rejected" : "left" });
  }));
  const rej = await realm.me.rejectInvitation({ tenantId: "t one", userBearer: "u" });
  const left = await realm.me.leave({ tenantId: "t1", userBearer: "u" });
  assert.equal(rej.status, "rejected");
  assert.equal(left.status, "left");
  // A tenant id is a path SEGMENT and must be escaped, not interpolated raw.
  assert.match(hits[0]!, /\/me\/invitations\/t%20one\/reject$/);
  assert.match(hits[1]!, /\/me\/memberships\/t1\/leave$/);
});

test("me.chooseTenant: owner_cannot_be_revoked keeps its specific code", async () => {
  const realm = realmWith(mkFetch(() => json({
    error: "transfer ownership first", code: "owner_cannot_be_revoked",
  }, 409)));
  await assert.rejects(
    () => realm.me.chooseTenant({ tenantId: "t1", userBearer: "u" }),
    (err: unknown) => {
      assert.ok(err instanceof RealmError);
      // Must NOT collapse into the generic 409 `conflict` — the remedy
      // (transfer ownership) is specific to this code.
      assert.equal(err.code, "owner_cannot_be_revoked");
      assert.equal(err.httpStatus, 409);
      return true;
    },
  );
});

test("auth.login: decodes the ADR-092 picker alongside real tokens", async () => {
  const fetch = (async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input.toString();
    const body = typeof init?.body === "string" ? JSON.parse(init.body) : {};
    if (url.endsWith("/auth/login") && body.grant_type === "platform_api_key") {
      return json({ subject_type: "platform", access_token: "pt_x", refresh_token: "rt", expires_in: 300 });
    }
    return json({
      access_token: "atok", refresh_token: "rtok", expires_in: 900,
      user: { id: "u1" }, tenants: [],
      tenant_choice_required: true,
      tenant_choices: [
        { tenant_id: "t1", display_name: "Acme", is_owner: false },
        { tenant_id: "t2", display_name: "Mine", is_owner: true },
      ],
    });
  }) as typeof fetch;
  const realm = realmWith(fetch);
  const out = await realm.auth.login({ method: "google", providerToken: "tok" });
  // The picker is a prompt, not an auth failure: the login still mints.
  assert.equal(out.accessToken, "atok");
  assert.equal(out.tenantChoiceRequired, true);
  assert.deepEqual(out.tenantChoices, [
    { tenantId: "t1", displayName: "Acme", isOwner: false },
    // An owned membership cannot be given up — do not offer it.
    { tenantId: "t2", displayName: "Mine", isOwner: true },
  ]);
});

test("auth.login: an ordinary login grows no picker fields", async () => {
  const fetch = (async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input.toString();
    const body = typeof init?.body === "string" ? JSON.parse(init.body) : {};
    if (url.endsWith("/auth/login") && body.grant_type === "platform_api_key") {
      return json({ subject_type: "platform", access_token: "pt_x", refresh_token: "rt", expires_in: 300 });
    }
    return json({
      access_token: "atok", refresh_token: "rtok", expires_in: 900,
      user: { id: "u1" }, tenants: [],
    });
  }) as typeof fetch;
  const out = await realmWith(fetch).auth.login({ method: "google", providerToken: "tok" });
  assert.equal(out.tenantChoiceRequired, undefined);
  assert.equal(out.tenantChoices, undefined);
});

test("config.get: pending reconciliation rides BESIDE config, never inside it", async () => {
  const realm = realmWith(mkFetch(() => json({
    id: "r",
    config: { single_tenant_membership: true },
    single_tenant_pending_reconciliation: 3,
  })));
  const out = await realm.config.get();
  assert.equal(out.config["single_tenant_membership"], true);
  assert.equal(out.singleTenantPendingReconciliation, 3);
  assert.equal("single_tenant_pending_reconciliation" in out.config, false);
});

test("config.get: pending reconciliation is undefined when the rule is off", async () => {
  const realm = realmWith(mkFetch(() => json({
    id: "r", config: { single_tenant_membership: false },
  })));
  const out = await realm.config.get();
  // undefined, not 0: "not reported" and "reported as drained" are different
  // facts, and a caller rendering the number must tell them apart.
  assert.equal(out.singleTenantPendingReconciliation, undefined);
});

test("me.acceptInvitation: own route, no body, escaped tenant id", async () => {
  const hits: string[] = [];
  const realm = realmWith(mkFetch((req) => {
    hits.push(req.url);
    assert.equal(req.body, undefined, "this route takes no body");
    return json({ tenant_id: "t one", status: "accepted" });
  }));
  const acc = await realm.me.acceptInvitation({ tenantId: "t one", userBearer: "u" });
  assert.equal(acc.status, "accepted");
  // The route is /accept, NOT /reject with a flag: the two are separate issuer
  // endpoints (ADR-095 D5), and pointing accept at reject's path would decline
  // the invitation it was asked to take. A tenant id is a path SEGMENT and must
  // be escaped, not interpolated raw.
  assert.match(hits[0]!, /\/me\/invitations\/t%20one\/accept$/);
});

test("me.acceptInvitation: not_pending keeps its specific code", async () => {
  const realm = realmWith(mkFetch(() => json({
    error: "invitation already answered", code: "not_pending",
  }, 409)));
  await assert.rejects(
    () => realm.me.acceptInvitation({ tenantId: "t1", userBearer: "u" }),
    (err: unknown) => {
      assert.ok(err instanceof RealmError);
      // Must NOT collapse into the generic 409 `conflict`: `not_pending`
      // (already answered) and `not_invited` (already a member) have different
      // remedies, and only the code tells them apart.
      assert.equal(err.code, "not_pending");
      assert.equal(err.httpStatus, 409);
      return true;
    },
  );
});
